File: Python/BOJ_2630_색종이 만들기.py
import sys

def cut(x, y, length, arr):
    global white_paper, blue_paper
    start = arr[x][y]; count = 0
    for i in range(length):
        for j in range(length):
            if arr[x + i][y + j] != start:
                count += 1
            if count != 0:
                break
        if count != 0:
            break
    
    if count == 0:
        if start == 0:
            white_paper += 1
        else:
            blue_paper += 1
    else:
        cut(x, y, length // 2, arr)
        cut(x, y + (length // 2), length // 2, arr)
        cut(x + (length // 2), y, length // 2, arr)
        cut(x + (length // 2), y + (length // 2), length // 2, arr)

size = int(sys.stdin.readline())
table = []
for i in range(size):
    table.append(list(map(int, sys.stdin.readline().split())))

white_paper = 0; blue_paper = 0
cut(0, 0, size, table)
print("%d\n%d" %(white_paper, blue_paper))

File: C++/BOJ_2630_색종이 만들기.cpp
#include <iostream>
#include <vector>

using namespace std;

void cut(int x, int y, int size, vector <vector <int>>& table, int& white, int& blue)
{
    int start, count;
    
    start = table[x][y];
    count = 0;
    
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            if (table[x + i][y + j] != start) {
                count++;
            }
            if (count) break;
        }
        if (count) break;
    }
    
    if (!count) {
        if (start == 0) white++;
        else blue++;
    } else {
        cut(x, y, size / 2, table, white, blue);
        cut(x, y + (size / 2), size / 2, table, white, blue);
        cut(x + (size / 2), y, size / 2, table, white, blue);
        cut(x + (size / 2), y + (size / 2), size / 2, table, white, blue);
    }
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    
    int size, white, blue;
    vector <vector <int>> table;
    
    cin >> size;
    for (int i = 0; i < size; i++) {
        table.emplace_back(vector <int> (size, 0));
        for (int j = 0; j < size; j++) {
            cin >> table[i][j];
        }
    }
    
    white = 0; blue = 0;
    cut(0, 0, size, table, white, blue);
    cout << white << '\n' << blue << '\n';

    return 0;
}
